File: 0948-sort-an-array/0948-sort-an-array.java
class Solution {
    public int[] sortArray(int[] nums) {
        mergeSort(nums);
        return nums;
    }


    private static void mergeSort(int[] array) {
        int size = array.length;

        if (size < 2) {
            return;
        }

        int middle = size/2;

        int[] leftArray = new int[middle];
        int[] rightArray = new int[size-middle];

        for (int i = 0; i < middle; i++) {
            leftArray[i] = array[i];
        }

        for (int j = middle; j < size; j++) {
            rightArray[j-middle] = array[j];
        }

        mergeSort(leftArray);
        mergeSort(rightArray);

        merge(array, leftArray, rightArray);
    }


    private static void merge(int[] array, int[] leftArray, int[] rightArray) {
        int leftSize = leftArray.length;
        int rightSize = rightArray.length;

        int i = 0, j = 0, k = 0;
        while (i < leftSize && j < rightSize) {
            if (leftArray[i] <= rightArray[j]) {
                array[k] = leftArray[i];
                i++;
            }
            else {
                array[k] = rightArray[j];
                j++;
            }
            k++;
        }

        while (i < leftSize) {
            array[k] = leftArray[i];
            i++;
            k++;
        }

        while (j < rightSize) {
            array[k] = rightArray[j];
            j++;
            k++;
        }
    }
}

File: 0948-sort-an-array/0948-sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        merge_sort(nums)
        return nums

def merge_sort(arr):
    size = len(arr)

    if size < 2:
        return

    middle = size//2

    left_half = arr[:middle]
    right_half = arr[middle:]

    merge_sort(left_half)
    merge_sort(right_half)

    merge(arr, left_half, right_half)

def merge(arr, left, right):
    left_size = len(left)
    right_size = len(right)

    i = j = k = 0
    
    while (i < left_size and j < right_size):
        if (left[i] <= right[j]):
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while (i < left_size):
        arr[k] = left[i]
        i += 1
        k += 1
    
    while (j < right_size):
        arr[k] = right[j]
        j += 1
        k += 1
    


